pre signed url is not accessible after 15 mins, throws clock skew error

### Environment info
- NooBaa Version: **noobaa-core-5.17.0-20241023.el9.x86_64**
- Platform: **standalone noobaa**

### Actual behavior
pre signed url creation is success and accessible via browser as well as curl command.
However after 15 mins (approximately) , the url becomes inaccessible and gives clockskew error.

### Expected behavior
If I set the expiry time of the pre signed url, the url must be accessible till that expiry time.

### Steps to reproduce
1. Upload a text file with some readable data in to bucket . 
[root@redshirt1 ~]# cat test123.txt 
abcd
[root@redshirt1 ~]# aws-test s3 cp test123.txt s3://bucket1001

2. Generate a pre signed url with expiration time of 7 days
[root@redshirt1 ~]# aws-test s3 presign s3://bucket1001/test123.txt --expires-in 604800

3. Now try to read the url via curl command or browser , Getting read properly.
Noobaa.log shows the following entry

```
Oct 30 02:45:36 enfield-x-master [1997121]: [nsfs/1997121]   [L0] core.sdk.bucketspace_fs:: BucketSpaceFS.read_bucket_sdk_info: bucket name bucket1001
Oct 30 02:45:36 enfield-x-master [1997121]: [nsfs/1997121]   [L0] core.sdk.bucketspace_fs:: BucketSpaceFS.read_bucket_sdk_info: bucket name bucket1001
Oct 30 02:45:36 enfield-x-master [1997121]: [nsfs/1997121]   [L0] core.sdk.namespace_fs:: NamespaceFS: read_object_stream { file_path: '/ibm/fs1/account1001/bucket1001/test123.txt', start: 0, end: Infinity, size: 5 }
Oct 30 02:45:36 enfield-x-master [1997121]: [nsfs/1997121]   [L0] core.sdk.namespace_fs:: NamespaceFS: read_object_stream completed file /ibm/fs1/account1001/bucket1001/test123.txt { num_bytes: 5, num_buffers: 1, avg_buffer: 5, log2_size_histogram: { '3': 1 } }

```

4. After roughly 15 mins do the same operations in step3 again.

`Oct 30 03:08:47 enfield-x-master [1997120]: [nsfs/1997120] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>RequestTimeTooSkewed</Code><Message>The difference between the request time and the server&apos;s time is too large.</Message><Resource>/bucket1001/test123.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=CoQUAQeCnOBNmROQxmuf%2F20241030%2Fus-east-1%2Fs3%2Faws4_request&amp;X-Amz-Date=20241030T094528Z&amp;X-Amz-Expires=604800&amp;X-Amz-SignedHeaders=host&amp;X-Amz-Signature=e09fada28fccfbb058cf235d45758bb119a4d5ffaa317cc39f5ae137d85951a4</Resource><RequestId>m2vps331-2q0kh1-ndv</RequestId></Error> GET /bucket1001/test123.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=CoQUAQeCnOBNmROQxmuf%2F20241030%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241030T094528Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=e09fada28fccfbb058cf235d45758bb119a4d5ffaa317cc39f5ae137d85951a4 {"host":"9.30.217.60:6443","connection":"keep-alive","cache-control":"max-age=0","sec-ch-ua":"\"Chromium\";v=\"130\", \"Google Chrome\";v=\"130\", \"Not?A_Brand\";v=\"99\"","sec-ch-ua-mobile":"?0","sec-ch-ua-platform":"\"macOS\"","upgrade-insecure-requests":"1","user-agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7","sec-fetch-site":"none","sec-fetch-mode":"navigate","sec-fetch-user":"?1","sec-fetch-dest":"document","accept-encoding":"gzip, deflate, br, zstd","accept-language":"en-US,en;q=0.9,ms;q=0.8,ml;q=0.7"} Error: The difference between the request time and the server's time is too large.    at Object.check_headers (/usr/local/noobaa-core/src/util/http_utils.js:602:15)    at handle_request (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:106:16)    at Object.s3_rest [as handler] (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:66:15)    at Server.endpoint_request_handler (/usr/local/noobaa-core/src/endpoint/endpoint.js:286:28)    at Server.emit (node:events:518:28)    at Server.emit (node:domain:488:12)    at parserOnIncoming (node:_http_server:1143:12)    at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)`


### More information - Screenshots / Logs / Other output

--- a/src/endpoint/s3/s3_errors.js
+++ b/src/endpoint/s3/s3_errors.js
@@ -534,6 +534,21 @@ S3Error.InvalidEncodingType = Object.freeze({
     message: 'Invalid Encoding Method specified in Request',
     http_code: 400,
 });
+S3Error.AuthorizationQueryParametersError = Object.freeze({
+    code: 'AuthorizationQueryParametersError',
+    message: 'X-Amz-Expires must be less than a week (in seconds); that is, the given X-Amz-Expires must be less than 604800 seconds',
+    http_code: 400,
+});
+S3Error.RequestExpired = Object.freeze({
+    code: 'AccessDenied',
+    message: 'Request has expired',
+    http_code: 403,
+});
+S3Error.RequestNotValidYet = Object.freeze({
+    code: 'AccessDenied',
+    message: 'request is not valid yet',
+    http_code: 403,
+});
 
 ////////////////////////////////////////////////////////////////
 // S3 Select                                                  //

--- a/src/util/http_utils.js
+++ b/src/util/http_utils.js
@@ -597,8 +597,15 @@ function check_headers(req, options) {
     if (isNaN(req_time) && !req.query.Expires && is_not_anonymous_req) {
         throw new options.ErrorClass(options.error_access_denied);
     }
-
-    if (Math.abs(Date.now() - req_time) > config.AMZ_DATE_MAX_TIME_SKEW_MILLIS) {
+    // futureus presigned url request should throw AccessDenied with request is no valid yet message
+    // we add a grace period of one second
+    const is_presigned_url = req.query.Expires || (req.query['X-Amz-Date'] && req.query['X-Amz-Expires']);
+    if (is_presigned_url && (req_time > (Date.now() + 2000))) {
+        throw new S3Error(S3Error.RequestNotValidYet);
+    }
+    // on regular requests the skew limit is 15 minutes
+    // on presigned url requests we don't need to check skew
+    if (!is_presigned_url && (Math.abs(Date.now() - req_time) > config.AMZ_DATE_MAX_TIME_SKEW_MILLIS)) {
         throw new options.ErrorClass(options.error_request_time_too_skewed);
     }
 }

--- a/src/util/signature_utils.js
+++ b/src/util/signature_utils.js
@@ -9,6 +9,7 @@ const path = require('path');
 const crypto = require('crypto');
 const S3Error = require('../endpoint/s3/s3_errors').S3Error;
 const http_utils = require('./http_utils');
+const time_utils = require('./time_utils');
 const { RpcError } = require('../rpc');
 
 
@@ -126,9 +127,9 @@ function _string_to_sign_v4(req, signed_headers, xamzdate, region, service) {
 
 function _check_expiry_query_v4(request_date, expires_seconds) {
     const now = Date.now();
-    const expires = (new Date(request_date).getTime()) + (Number(expires_seconds) * 1000);
+    const expires = (new Date(time_utils.parse_amz_date(request_date)).getTime()) + (Number(expires_seconds) * 1000);
     if (now > expires) {
-        throw new Error('Authentication Expired (V4)');
+        throw new S3Error(S3Error.RequestExpired);
     }
 }
 
@@ -180,7 +181,7 @@ function _check_expiry_query_s3(expires_epoch) {
     const now = Date.now();
     const expires = Number(expires_epoch) * 1000;
     if (now > expires) {
-        throw new Error('Authentication Expired (S3)');
+        throw new S3Error(S3Error.RequestExpired);
     }
 }
 
@@ -285,12 +286,20 @@ function make_auth_token_from_request(req) {
  */
 function check_request_expiry(req) {
     if (req.query['X-Amz-Date'] && req.query['X-Amz-Expires']) {
+        _check_expiry_limit(req.query['X-Amz-Expires']);
         _check_expiry_query_v4(req.query['X-Amz-Date'], req.query['X-Amz-Expires']);
     } else if (req.query.Expires) {
+        _check_expiry_limit(req.query.Expires);
         _check_expiry_query_s3(req.query.Expires);
     }
 }
 
+// expiry_seconds limit is 7 days
+function _check_expiry_limit(expiry_seconds) {
+    if (Number(expiry_seconds) > 7 * 24 * 60 * 60 * 1000) {
+        throw new S3Error(S3Error.AuthorizationQueryParametersError);
+    }
+}
 
 /**
  *

